tgui/packages/tgui/components/MenuBarDropdown.tsx

```diff
@@ -0,0 +1,208 @@
+};
+
+class Menu extends Component<MenuProps> {
+  private readonly handleClick: (event) => void;
+
+  constructor(props) {
+    super(props);
+    this.handleClick = (event) => {
+      if (!this.props.menuRef.current) {
+        logger.log(`Menu.handleClick(): No ref`);
+        return;
+      }
+
+      if (this.props.menuRef.current.contains(event.target)) {
+        logger.log(`Menu.handleClick(): Inside`);
+      } else {
+        logger.log(`Menu.handleClick(): Outside`);
+        this.props.onOutsideClick();
+      }
+    };
+  }
+
+  // eslint-disable-next-line react/no-deprecated
+  componentWillMount() {
+    window.addEventListener('click', this.handleClick);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('click', this.handleClick);
+  }
+
+  render() {
+    const { open, options, width } = this.props;
+    const menu = open ? (
+      <div
+        className={'MenuBar__menu'}
+        style={{
+          width: width,
+        }}>
+        {options.length ? options : 'No Options Found'}
+      </div>
+    ) : null;
+
+    return { menu };
+  }
+}
+
+type MenuBarDropdownProps = {
+  open: boolean;
+  openWidth: string;
+  options: InfernoNodeArray;
+  disabled?: boolean;
+  display: InfernoNode;
+  onMouseOver: () => void;
+  onClick: () => void;
+  onOutsideClick: () => void;
+  className?: string;
+};
+
+class MenuBarButton extends Component<MenuBarDropdownProps> {
+  private readonly menuRef: RefObject<HTMLDivElement>;
+
+  constructor(props) {
+    super(props);
+    this.menuRef = createRef();
+  }
+
+  render() {
+    const { props } = this;
+    const {
+      open,
+      openWidth,
+      options,
+      disabled,
+      display,
+      onMouseOver,
+      onClick,
+      onOutsideClick,
+      ...boxProps
+    } = props;
+    const { className, ...rest } = boxProps;
+
+    return (
+      <div ref={this.menuRef} className="MenuBar">
+        <Box
+          className={classes(['MenuBar__MenuBarButton', className])}
+          {...rest}
+          onClick={disabled ? undefined : onClick}
+          onmouseover={onMouseOver}>
+          <span className="MenuBar__selected-text">{display}</span>
+        </Box>
+        {open && (
+          <Menu
+            open={open}
+            width={openWidth}
+            options={options}
+            menuRef={this.menuRef}
+            onOutsideClick={onOutsideClick}
+          />
+        )}
+      </div>
+    );
+  }
+}
+
+type MenuBarItemProps = {
+  entry: string;
+  openWidth: string;
+  display: InfernoNode;
+  options: InfernoNodeArray;
+  setOpenMenuBar: (entry: string | null) => void;
+  openMenuBar: string | null;
+  setOpenOnHover: (flag: boolean) => void;
+  openOnHover: boolean;
+  disabled?: boolean;
+  className?: string;
+};
+
+export const MenuBarDropdown = (props: MenuBarItemProps) => {
+  const {
+    entry,
+    openWidth,
+    display,
+    options,
+    setOpenMenuBar,
+    openMenuBar,
+    setOpenOnHover,
+    openOnHover,
```

Comment:
These 4 properties (+ entry) leak implementation details of your UI and don't seem to belong here, you need to abstract it better. If I had to use this somewhere else, I'd have no idea how to use it. Maybe all this display logic should be implemented in the UI itself, and here only expose "open" and "onOpen/onClose"?